crml/staking/src/rewards/mod.rs

```diff
@@ -151,17 +167,49 @@ where
 				Self::calculate_npos_payouts(&validator, *validator_commission, stake_map, total_payout_share)
 			})
 			.for_each(|(account, payout)| {
-				total_payout_imbalance.maybe_subsume(T::CurrencyToReward::deposit_into_existing(&account, payout).ok());
+				EraRemainedPayouts::<T>::mutate(|p| p.push((account, payout)));
 			});
 
-		// Any unallocated reward amount can go to the development fund
-		let remainder = total_payout.saturating_sub(total_payout_imbalance.peek());
-		total_payout_imbalance.subsume(
-			T::CurrencyToReward::deposit_into_existing(&T::TreasuryModuleId::get().into_account(), remainder)
-				.unwrap_or_else(|_| PositiveImbalanceOf::<T>::zero()),
-		);
+		EraRemainedRewardAmount::<T>::put(total_payout.saturating_sub(total_payout_imbalance.peek()));
+	}
+
+	fn process_reward_payouts(remained_blocks: Self::BlockNumber) -> Weight {
+		let remained_payouts = EraRemainedPayouts::<T>::get().len();
+		let quota = Self::calculate_payout_quota(remained_payouts, remained_blocks);
+		if quota == 0 {
+			return T::WeightInfo::process_zero_payouts();
+		}
+
+		let weight = T::WeightInfo::process_reward_payouts(quota);
+
+		EraRemainedPayouts::<T>::mutate(|p| {
```

Comment:
this will decode the entire EraRemainedPayouts storage,
not sure how much affect it will have though...
other option I can think of is to pre-bucket the payouts at end of an era, then you only need to read 1 bucket per payout

Reply:
The good thing is with benchmarking, we can try different approaches and test their actual cost.

Reply:
I would suggest we postpone this to later when the actual usage and scale of our network is better known. That's to avoid any premature optimisations.